### ISSUE TITLE
Applications exit/crash on window resize on Linux systems

### ISSUE THREAD

**Issue by schnippl0r** [parasyte/pixels#121]
Hi @parasyte,

I have written a small application using pixels and found out that the application crashes on a window resize. I can reproduce the same behaviour with running the example application `minimal-winit`.

Resizing the window provides the following error message:
`[2020-10-09T18:56:49Z ERROR minimal_winit] pixels.render() failed: The GPU failed to acquire a swapchain frame.`

Removing the error handling on the `pixels.render()` call (line 49) and just `unwrap()`-ing it delivers the following error message:
`thread 'main' panicked at 'called 'Result::unwrap()' on an 'Err' value: Swapchain(Outdated)', examples/minimal-winit/src/main.rs:49:29`

I tested the window resize on different Linux systems and graphics drivers: ArchLinux+amdgpu, ArchLinux+nvidia, Debian 10+intel. All systems were running X11, not Wayland.
Performing a window resize with the example running under Windows 10 is working as expected.

I have also tested other examples from frameworks that base on winit (piston, amethyst) and window resizing works as expected.

Kind regards

Full Backtrace:
```
thread 'main' panicked at 'called 'Result::unwrap()' on an 'Err' value: Swapchain(Outdated)', examples/minimal-winit/src/main.rs:49:29
stack backtrace:
   0:     0x562710bc8340 - std::backtrace_rs::backtrace::libunwind::trace::h577ea05e9ca4629a
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/../../backtrace/src/backtrace/libunwind.rs:96
   1:     0x562710bc8340 - std::backtrace_rs::backtrace::trace_unsynchronized::h50b9b72b84c7dd56
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/../../backtrace/src/backtrace/mod.rs:66
   2:     0x562710bc8340 - std::sys_common::backtrace::_print_fmt::h6541cf9823837fac
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:79
   3:     0x562710bc8340 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf64fbff071026df5
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:58
   4:     0x562710bead6c - core::fmt::write::h9ddafa4860d8adff
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/core/src/fmt/mod.rs:1082
   5:     0x562710bc57a7 - std::io::Write::write_fmt::h1d2ee292d2b65481
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/io/mod.rs:1514
   6:     0x562710bca670 - std::sys_common::backtrace::_print::ha25f9ff5080d886d
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:61
   7:     0x562710bca670 - std::sys_common::backtrace::print::h213e8aa8dc5405c0
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:48
   8:     0x562710bca670 - std::panicking::default_hook::{{closure}}::h6482fae49ef9d963
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:200
   9:     0x562710bca3bc - std::panicking::default_hook::he30ad7589e0970f9
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:219
  10:     0x562710bcacd3 - std::panicking::rust_panic_with_hook::haa1ed36ada4ffb03
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:569
  11:     0x562710bca8a9 - std::panicking::begin_panic_handler::{{closure}}::h7001af1bb21aeaeb
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:476
  12:     0x562710bc87cc - std::sys_common::backtrace::__rust_end_short_backtrace::h39910f557f5f2367
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/sys_common/backtrace.rs:153
  13:     0x562710bca869 - rust_begin_unwind
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:475
  14:     0x562710be8901 - core::panicking::panic_fmt::h4e2659771ebc78eb
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/core/src/panicking.rs:85
  15:     0x562710be8723 - core::option::expect_none_failed::h448b58a024c2c33a
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/core/src/option.rs:1221
  16:     0x56270ffa17dd - core::result::Result<T,E>::unwrap::h23db5a3d825cf25a
                               at /home/_/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:973
  17:     0x56270ffb3fba - minimal_winit::main::{{closure}}::h06100a03ce68d0d7
                               at /tmp/pixels/examples/minimal-winit/src/main.rs:49
  18:     0x56270ffb1163 - winit::platform_impl::platform::sticky_exit_callback::h1516e1425b3c5c8a
                               at /home/_/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/mod.rs:698
  19:     0x56270ffb22a5 - winit::platform_impl::platform::x11::EventLoop<T>::run_return::h3cf3834f72dcc8c4
                               at /home/_/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/x11/mod.rs:312
  20:     0x56270ffb328b - winit::platform_impl::platform::x11::EventLoop<T>::run::h157996e81a5c9084
                               at /home/_/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/x11/mod.rs:390
  21:     0x56270ffb0d18 - winit::platform_impl::platform::EventLoop<T>::run::h6739777b3595d3a7
                               at /home/_/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/platform_impl/linux/mod.rs:645
  22:     0x56270ffbfd33 - winit::event_loop::EventLoop<T>::run::h98885696453ed85f
                               at /home/_/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.2/src/event_loop.rs:149
  23:     0x56270ffd69bc - minimal_winit::main::ha83f5fbf0b8052eb
                               at /tmp/pixels/examples/minimal-winit/src/main.rs:45
  24:     0x56270ffd956b - core::ops::function::FnOnce::call_once::hb4b1578fcdd13503
                               at /home/_/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  25:     0x56270ff8d19e - std::sys_common::backtrace::__rust_begin_short_backtrace::h0505aa401a658e29
                               at /home/_/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137
  26:     0x56270ffbe681 - std::rt::lang_start::{{closure}}::h7f28f9e41a3d708b
                               at /home/_/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:66
  27:     0x562710bcb0d1 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h6a3209f124be2235
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/core/src/ops/function.rs:259
  28:     0x562710bcb0d1 - std::panicking::try::do_call::h88ce358792b64df0
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:373
  29:     0x562710bcb0d1 - std::panicking::try::h6311c259678e50fc
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panicking.rs:337
  30:     0x562710bcb0d1 - std::panic::catch_unwind::h56c5716807d659a1
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/panic.rs:379
  31:     0x562710bcb0d1 - std::rt::lang_start_internal::h73711f37ecfcb277
                               at /rustc/18bf6b4f01a6feaf7259ba7cdae58031af1b7b39/library/std/src/rt.rs:51
  32:     0x56270ffbe657 - std::rt::lang_start::h715b996db0bf322c
                               at /home/_/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:65
  33:     0x56270ffd70ba - main
  34:     0x7fbaba1d2152 - __libc_start_main
  35:     0x56270ff8d0be - _start
  36:                0x0 - <unknown>
```

**Comment by bjorn3** [comment]
Providing your code may help finding the problem.

**Comment by schnippl0r** [comment]
> Providing your code may help finding the problem.

You can just use the `minimal-winit` example.

**Comment by parasyte** [comment]
This is pretty unusual. The swapchain is recreated by the resize handler: https://github.com/parasyte/pixels/blob/e5f92519f17026a9f7e4821c5e9fc72895cdc857/src/lib.rs#L220-L230

This could actually be a race condition between resize and render. The event loop in the example uses `winit_input_helper = "0.6.0"` which consumes the `Event` type when updating the state. This forces our hand to handle the render before resize. The latest version of this crate takes `&Event`, so we can actually reorder these.

@schnippl0r Try this patch, let me know if it helps:

```patch
diff --git a/examples/minimal-winit/Cargo.toml b/examples/minimal-winit/Cargo.toml
index 7acbc4e..758eddf 100644
--- a/examples/minimal-winit/Cargo.toml
+++ b/examples/minimal-winit/Cargo.toml
@@ -13,5 +13,5 @@ default = ["optimize"]
 env_logger = "0.7.1"
 log = "0.4.8"
 pixels = { path = "../.." }
-winit = "0.22.0"
-winit_input_helper = "0.6.0"
+winit = "0.23.0"
+winit_input_helper = "0.8.0"
diff --git a/examples/minimal-winit/src/main.rs b/examples/minimal-winit/src/main.rs
index fcaeff7..d1ecdf0 100644
--- a/examples/minimal-winit/src/main.rs
+++ b/examples/minimal-winit/src/main.rs
@@ -43,21 +43,8 @@ fn main() -> Result<(), Error> {
     let mut world = World::new();
 
     event_loop.run(move |event, _, control_flow| {
-        // Draw the current frame
-        if let Event::RedrawRequested(_) = event {
-            world.draw(pixels.get_frame());
-            if pixels
-                .render()
-                .map_err(|e| error!("pixels.render() failed: {}", e))
-                .is_err()
-            {
-                *control_flow = ControlFlow::Exit;
-                return;
-            }
-        }
-
         // Handle input events
-        if input.update(event) {
+        if input.update(&event) {
             // Close events
             if input.key_pressed(VirtualKeyCode::Escape) || input.quit() {
                 *control_flow = ControlFlow::Exit;
@@ -73,6 +60,19 @@ fn main() -> Result<(), Error> {
             world.update();
             window.request_redraw();
         }
+
+        // Draw the current frame
+        if let Event::RedrawRequested(_) = event {
+            world.draw(pixels.get_frame());
+            if pixels
+                .render()
+                .map_err(|e| error!("pixels.render() failed: {}", e))
+                .is_err()
+            {
+                *control_flow = ControlFlow::Exit;
+                return;
+            }
+        }
     });
 }
 
```

**Comment by schnippl0r** [comment]
Sadly, the patch does not fix the issue. The error messages remain the same.

**Comment by parasyte** [comment]
I'll have to get a VM going to try a repro. I'm on macOS and it's all working correctly here.

**Comment by JMS55** [comment]
I can't reproduce this with the minimal-winit example on my linux machine.

**Comment by schnippl0r** [comment]
> I can't reproduce this with the minimal-winit example on my linux machine, Fedora 32 + Intel IGPU.

I downloaded and booted into a live Fedora 32: x86_64 DVD ISO. After installing `libXScrnSaver-devel` the `minimal-winit` example compiled and ran fine.

I can confirm that the window resizes as expected. But I witnessed another issue: After updating the `winit` dependency to 0.23 and `winit_input_helper` to 0.8 and adding a borrow to `event` at line 60 in `src/main.rs`, the window lost the possibility to resize at all. I even tried adding `.with_resizable(true)` after line 33 but resizing was still impossible. So it seems to me that this new issue may be a regression in the `winit` crate itself.

@JMS55 can you confirm my observations?

And can please someone running an X11 machine try to resize the window of the `minimal-winit` example and reproduce the crash?

**Comment by bjorn3** [comment]
On Debian with X11 I get

```
[2020-10-10T16:27:01Z ERROR minimal_winit] pixels.render() failed: The GPU failed to acquire a swapchain frame.
```

followed by the example exiting.

**Comment by JMS55** [comment]
I can confirm what @schnippl0r said about updating deps causing issues. However, I believe this is a `winit_input_helper` problem. A project I made with `winit 0.23` and `pixels` works fine on the same machine.

**Comment by svenstaro** [comment]
I'm on Arch Linux, Intel i915 and X11 and can very easily reproduce the crash when resizing the window with the `minimal-winit` example.

It does work sometimes but now always.

    [2020-10-27T03:47:49Z ERROR minimal_winit] pixels.render() failed: The GPU failed to acquire a swapchain frame.

The _sometimes_ part would indeed point to this problem resulting from a race condition.

**Comment by svenstaro** [comment]
Here's a fun discovery I had:
```diff
diff --git a/examples/minimal-winit/src/main.rs b/examples/minimal-winit/src/main.rs
index fcaeff7..fdfe9a7 100644
--- a/examples/minimal-winit/src/main.rs
+++ b/examples/minimal-winit/src/main.rs
@@ -65,9 +65,11 @@ fn main() -> Result<(), Error> {
             }
 
             // Resize the window
-            if let Some(size) = input.window_resized() {
-                pixels.resize(size.width, size.height);
-            }
+            // if let Some(size) = input.window_resized() {
+            //     pixels.resize(size.width, size.height);
+            // }
+            let size = window.inner_size();
+            pixels.resize(size.width, size.height);
 
             // Update internal state and request a redraw
             world.update();
```

This appears to work just fine.

**Comment by svenstaro** [comment]
This appears to be related: https://github.com/gfx-rs/wgpu-rs/issues/352, especially https://github.com/gfx-rs/wgpu-rs/issues/352#issuecomment-656985476

**Comment by svenstaro** [comment]
I found a stupid workaround that works well enough for me:

```rust
    let mut resize_count = 0;

    event_loop.run(move |event, _, control_flow| {
        if let Event::RedrawRequested(_) = event {
            if resize_count > 0 {
                let size = window.inner_size();
                pixels.resize(size.width, size.height);
                resize_count -= 1;
            }

            let frame = pixels.get_frame();
            for pixel in frame.chunks_exact_mut(4) {
                pixel[0] = rng.gen();
                pixel[1] = rng.gen();
                pixel[2] = rng.gen();
                pixel[3] = 0xff;
            }
            if pixels
                .render()
                .map_err(|e| eprintln!("pixels.render() failed: {}", e))
                .is_err()
            {
                *control_flow = ControlFlow::Exit;
                return;
            }
        }

        if input.update(&event) {
            if input.key_pressed(VirtualKeyCode::Escape) || input.quit() {
                *control_flow = ControlFlow::Exit;
                return;
            }

            if let Some(size) = input.window_resized() {
                // Resize many times to work around this bug: https://github.com/parasyte/pixels/issues/121
                resize_count = 100;
                pixels.resize(size.width, size.height);
            }

            window.request_redraw();
        }
    });
```

Of course, this makes me deeply unhappy, but what can I do?!

**Comment by svenstaro** [comment]
The fix for wgpu-rs examples was [this](https://github.com/gfx-rs/wgpu-rs/pull/434). Perhaps we can do something similar here.

**Comment by svenstaro** [comment]
Ok, so I managed to produce what I consider a proper fix. Please have a look.

**Comment by schnippl0r** [comment]
The fix provided by @svenstaro fixes the issue for me. Thank you, good job!

**Comment by parasyte** [comment]
> The fix for wgpu-rs examples was [this](https://github.com/gfx-rs/wgpu-rs/pull/434). Perhaps we can do something similar here.

The swapchain is already created when the `SurfaceTexture` is resized. This was mentioned in a comment above.

I saw the patch in #122 but I'm wondering if this is an acceptable solution instead: https://github.com/gfx-rs/wgpu-rs/issues/352#issuecomment-657793198 I would really want to avoid recreating the swapchain while attempting to get the next frame.

**Comment by parasyte** [comment]
What I mean is that it would be the application's responsibility to pass the correct size to `pixels`. Something like this in `minimal-winit/src/main.rs`:

```rust
            // Resize the window
            if let Some(_) = input.window_resized() {
                let size = window.inner_size();
                pixels.resize(size.width, size.height);
            }
```

**Comment by joseluis** [comment]
The error is the exactly same as when all the resizing code is commented out. And resizing without waiting for `window_resized()` just works. Maybe it's related with `window_resized()` latency.

**Comment by parasyte** [comment]
Any latency with that method was already described, and a patch provided: https://github.com/parasyte/pixels/issues/121#issuecomment-706379944

**Comment by vrmiguel** [comment]
I'm also having this issue when using the `minimal-winit` example on Arch Linux + Radeon graphics.

**Comment by parasyte** [comment]
This is fixed by #136
